### ISSUE TITLE
How is NuGet.config consumed by dnu?

### ISSUE THREAD

**Issue by per-samuelsson** [aspnet/dnx#2975]
Tried experiementing with this, and I just cant seem to get it and neither find any relevant docs.

Basically, if I put a `NuGet.config` next to a `project.json`:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <clear />
    <add key="NuGet" value="https://nuget.org/api/v2/" />
    <add key="AspNetVNext" value="https://www.myget.org/F/aspnetrelease/api/v2" />
  </packageSources>
</configuration>
```

1. Should `dnu restore` consume those feeds?
2. Should `dnu feeds list` list them?

**Comment by davidfowl** [comment]
1. Yes
2. Yes

If you ran restore from that folder. A more typical file system layout is something like:

NuGet.config
src/Project1/project.json
src/Project2/project.json

**Comment by per-samuelsson** [comment]
This https://github.com/aspnet/dnx/issues/1635#issuecomment-92464398 indicates its only doable in local/personal `nuget.config`, not a "closer" one. Anyone can confirm this? Because this is basically my experience.

Why I'm after is the kind of chaining / overriding supported natively by NuGet, as described in [Chaining multiple configuration files](http://docs.nuget.org/consume/nuget-config-file#chaining-multiple-configuration-files). 

When reading [Configuring the feed used by dnu to restore packages](https://github.com/aspnet/Home/wiki/Configuring-the-feed-used-by-dnu-to-restore-packages), and in particular the section:

> In addition, the samples on the Home repo and MusicStore as well as Bug Tracker have NuGet.config in their root folders. These have the vnext feed in them and hence ‘dnu restore’ on these applications pulls in the right set of packages.

its hard not to get the impression this stuff should work as I'm trying to use it.

**Comment by per-samuelsson** [comment]
> Yes
> Yes

Thanks, @davidfowl. Then never mind my last comment (wrote it while you answered :smile:). Then it's something that just don't work on my env then. Now I know it **should** work like that, I'll try some more. It might be the dreaded stuff with bad XML-format, for example.

> A more typical file system layout is something like [...]

Thats exactly the kind of thing I tried set up first. Just simplified it more for describing the issue.

**Comment by davidfowl** [comment]
@per-samuelsson I know it works because we use it in our own repository layout :smile:

**Comment by per-samuelsson** [comment]
Wow. Turned out to be one of those really freaky little things you luckily dont hit to often.

**TL;DR**: a UTF-8 encoded `NuGet.config` with some hidden byte was not correctly parsed, and hence just ignored (which is an old, known NuGet story, e.g #2152).

If someone is masochistic enough to recreate the issue, for whatever reason, here's how:

1. Create a new empty file in some editor supporting UTF-8. I used Notepad++. Name the file `NuGet.config` and make sure you set **Encoding -> Encode in UTF-8**.
2. CTRL+C [the MusicStore config](https://github.com/aspnet/MusicStore/blob/master/NuGet.Config) and paste it into your new file. Save it.
3. Run `dnx feeds list` in a setup as described in the OP. See it being ignored.

The funny (or rather: not so funny) thing is that *you can't see the wrong format with the eye*. It's an extra, non visible byte in line 1, column 1, just before the XML prolog `<?xml version="1.0" encoding="UTF-8"?>`. You can see it if you use the arrow keys - it takes two strokes to get to the inital `<`. 

How sick is that? :dizzy_face: 

Seems not even the recommendation to always use an editor that report parsing errors will always help you away from that mess.

> NuGet will silently ignore the entire configuration file if it encounters any XML parsing issues (such as mismatched begin/end nodes, invalid quotation, etc.); therefore an editor that reports parsing issues is recommended.

Conclusion: +1 for #2152.

**Comment by per-samuelsson** [comment]
Oh, and with the above being said, feel free to close this issue. Thanks for the help once again, @davidfowl!

**Comment by davidfowl** [comment]
Ugh what a horrible bug /cc @yishaigalatzer

**Comment by jordimontana82** [comment]
Jaysus ! I was going mad with the exact same issue.... thx a mill @yishaigalatzer !

**Comment by yishaigalatzer** [comment]
We will fix that at some point in the near future. It definitely sucks

**Comment by khellang** [comment]
Ah, the good old [BOM](https://en.wikipedia.org/wiki/Byte_order_mark). Lovely! :heart:

**Comment by jordimontana82** [comment]
If helps, the BOM can be removed easily using this command 

    sed -i '1 s/^\xef\xbb\xbf//' NuGet.config

I've experienced the same issue while moving shell scripts edited in Windows to docker containers in linux and the command is really handy.
